src/main/scala/mesosphere/util/state/zk/ZKStore.scala

```diff
@@ -0,0 +1,127 @@
+    * @return Some value, if the store operation is successful otherwise None
+    */
+  override def save(entity: PersistentEntity): Future[ZKEntity] = {
+    val zk = zkEntity(entity)
+    def update(nodeData: ZNode.Data): Future[ZKEntity] = {
+      val future: TWFuture[Data] = zk.node.setData(zk.data.toProto.toByteArray, nodeData.stat.getVersion)
+      mapFutureResult(future, s"Can not update entity $entity") { data =>
+        TWFuture.value(zk.copy(read = Some(data)))
+      }
+    }
+    def create(): Future[ZKEntity] = {
+      mapFutureResult(zk.node.create(zk.data.toProto.toByteArray), s"Can not update entity $entity") { data =>
+        zk.node.getData().map{ data => zk.copy(read = Some(data)) }
+      }
+    }
+    zk.read.map(update).getOrElse(create())
+  }
+
+  /**
+    * Delete an entry with given identifier.
+    */
+  override def delete(key: ID): Future[ZKEntity] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    mapFutureResult(node.exists(), s"Can not delete entity $key") { data =>
+      node.delete(data.stat.getVersion).map(n => ZKEntity(n, ZKData(key, UUID.randomUUID())))
+    }
+  }
+
+  override def allIds(): Future[Seq[ID]] = {
+    root.getChildren().map(_.children.map(_.name)).asScala
+  }
+
+  private def mapFutureResult[A, B](value: TWFuture[A], errorMessage: String)(fn: A => TWFuture[B]): Future[B] = {
+    value.liftToTry.flatMap {
+      case Return(data)               => fn(data)
+      case Throw(ex: KeeperException) => throw new StoreCommandFailedException(errorMessage, ex)
+      case Throw(ex)                  => throw ex
+    }.asScala
+  }
+
+  private def zkEntity(entity: PersistentEntity): ZKEntity = {
+    entity match {
+      case zk: ZKEntity => zk
+      case _            => throw new IllegalArgumentException("Can not handle this kind of entity")
+    }
+  }
+
+  private def createNode(node: ZNode): ZNode = {
+    Await.result(node.exists().liftToTry.flatMap {
+      case Return(data) => TWFuture.value(node)
+      case Throw(ex)    => node.create()
+    })
+  }
+}
+
+case class ZKEntity(node: ZNode, data: ZKData, read: Option[ZNode.Data] = None) extends PersistentEntity {
+  override def mutate(updated: Array[Byte]): PersistentEntity = copy(data = data.copy(bytes = updated))
+  override def bytes: Array[Byte] = data.bytes
+}
+
+case class ZKData(name: String, uuid: UUID, bytes: Array[Byte] = Array.empty) {
+  def toProto: Protos.ZKStoreEntry = Protos.ZKStoreEntry.newBuilder()
+    .setName(name)
+    .setUuid(ByteString.copyFromUtf8(uuid.toString))
+    .setValue(ByteString.copyFrom(bytes))
+    .build()
+}
+object ZKData {
+  def apply(bytes: Array[Byte]): ZKData = {
+    val proto = Protos.ZKStoreEntry.parseFrom(bytes)
+    new ZKData(proto.getName, UUIDUtil.uuid(proto.getUuid.toByteArray), proto.getValue.toByteArray)
+  }
+}
+
+object ZKStore {
+  implicit class Twitter2Scala[T](val twitterF: TWFuture[T]) extends AnyVal {
```

Comment:
I am somewhat an implicit conversion hater.

If I want to understand .asScala above, I need to check out the code and use my IDE. Otherwise I'll have a hard time finding the relevant code.